exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/TopN/TopNBatch.java

```diff
@@ -162,56 +173,67 @@ public void buildSchema() throws SchemaChangeException {
         return;
       case NONE:
         state = BatchState.DONE;
+      case EMIT:
+        throw new IllegalStateException("Unexpected EMIT outcome received in buildSchema phase");
       default:
-        return;
+        throw new IllegalStateException("Unexpected outcome received in buildSchema phase");
     }
   }
 
   @Override
   public IterOutcome innerNext() {
     recordCount = 0;
     if (state == BatchState.DONE) {
-      return IterOutcome.NONE;
+      return NONE;
     }
-    if (schema != null) {
-      if (getSelectionVector4().next()) {
+
+    // If both schema and priorityQueue are non-null and priority queue is not reset, that means we still have data
+    // to be sent downstream for the current record boundary
+    if (schema != null && priorityQueue != null && priorityQueue.isInitialized()) {
+      if (sv4.next()) {
         recordCount = sv4.getCount();
-        return IterOutcome.OK;
+        container.setRecordCount(recordCount);
       } else {
         recordCount = 0;
-        return IterOutcome.NONE;
+        container.setRecordCount(0);
       }
+      return getFinalOutcome();
     }
 
     try{
       outer: while (true) {
         Stopwatch watch = Stopwatch.createStarted();
-        IterOutcome upstream;
         if (first) {
-          upstream = IterOutcome.OK_NEW_SCHEMA;
+          laskKnownOutcome = IterOutcome.OK_NEW_SCHEMA;
```

Comment:
Thought this was a typo in one place until I realised this is how it is spelt all over. :)

Reply:
Fixed.